sympy/matrices/matrices.py

```diff
@@ -3888,59 +3891,164 @@ def LUdecomposition(self, iszerofunc=_iszero):
+        exactly once.
+        iszerofunc is a callable that returns a boolean indicating if its input
+        is zero.
+        If iszerofunc may return None if it cannot determine if its input is
+        zero.
+        The keyword argument rankcheck determines if this function raises a
+        ValueError when passed a matrix whose rank is strictly less than
+        min(num rows, num cols). The default behavior is to decompose a rank
+        deficient matrix. Pass rankcheck=True to raise a
+        ValueError instead. (This mimics the previous behavior of this function).
 
         See Also
         ========
 
         LUdecomposition
+        LUdecomposition_Simple
         LUdecompositionFF
         LUsolve
         """
-        if not self.is_square:
-            raise NonSquareMatrixError(
-                "A Matrix must be square to apply LUdecomposition_Simple().")
-        n = self.rows
-        A = self.as_mutable()
-        p = []
-        # factorization
-        for j in range(n):
-            for i in range(j):
-                for k in range(i):
-                    A[i, j] = A[i, j] - A[i, k] * A[k, j]
-            pivot = -1
-            for i in range(j, n):
-                for k in range(j):
-                    A[i, j] = A[i, j] - A[i, k] * A[k, j]
-                # find the first non-zero pivot, includes any expression
-                if pivot == -1 and not iszerofunc(A[i, j]):
-                    pivot = i
-            if pivot < 0:
-                # this result is based on iszerofunc's analysis of the possible pivots, so even though
-                # the element may not be strictly zero, the supplied iszerofunc's evaluation gave True
-                raise ValueError("No nonzero pivot found; inversion failed.")
-            if pivot != j:  # row must be swapped
-                A.row_swap(pivot, j)
-                p.append([pivot, j])
-            scale = 1 / A[j, j]
-            for i in range(j + 1, n):
-                A[i, j] = A[i, j] * scale
-        return A, p
+
+        if self.rows == 0 or self.cols == 0:
+            # Define LU decomposition of a matrix with no entries as a matrix
+            # of the same dimensions with all zero entries.
+            return self.zeros(self.rows, self.cols), []
+
+        lu = self.as_mutable()
+        row_swaps = []
+
+        pivot_col = 0
+        for pivot_row in range(0, lu.rows-1):
+
+            for candidate_pivot_col in range(pivot_col, self.cols):
+                for candidate_pivot_row in range(pivot_row, self.rows):
+                    iszeropivot = iszerofunc(lu[candidate_pivot_row,
+                                                candidate_pivot_col])
+                    if iszeropivot is None:
+                        iszeropivot = False
+                        break
+                    if not iszeropivot:
+                        break
+                if not iszeropivot:
+                    break
```

Comment:
You're doing a column by column search for a pivot, right?  I find the multiple negatives being used hard to follow.

There is a `_find_reasonable_pivot` function that you can use.  Maybe something like
```
for new_pivot_col in range(pivot_col, self.cols):
    pos, val, _, _ = _find_reasonable_pivot(lu[pivot_row:, new_pivot_col], iszerofunc)
    if pivot:
        new_pivot_row = pivot_row + pos
        break
```